python/CMakeLists.txt

```diff
@@ -56,6 +62,10 @@ nanobind_add_module(
   dolfinx/wrappers/petsc.cpp
   dolfinx/wrappers/refinement.cpp
 )
+# mpi4py and petsc4py use generated Cython headers
+# make sure to tell them that the limited API is in use
+target_compile_definitions(cpp PRIVATE CYTHON_COMPILING_IN_LIMITED_API=1)
```

Comment:
Should this only be enabled if `NANOBIND_SABI` is `STABLE_ABI`?

Reply:
I added the condition and a comment that it's actually a Cython bug that they are needed. If you want, we could leave that out and leave it to users' to add CYTHON_COMPILING_IN_LIMITED_API via CFLAGS.